Docs: Add Ingress documentation (#1552)
strimzi/strimzi-kafka-operator: documentation/book/con-kafka-listeners.adoc

* `route`
* `loadbalancer`
* `nodeport`
ifdef::Kubernetes[]
* `ingress`
endif::Kubernetes[]

> === Exposing Kafka using {OpenShiftName} Routes

Reviewer: Should Routes be `Routes` like it is in the following sentences?

For more information on using `Routes` to access Kafka, see xref:proc-accessing-kafka-using-routes-{context}[].

> === Exposing Kafka using loadbalancers

Reviewer: Should loadbalancers be `loadbalancer` like it is in the following sentences?

For more information on using node ports to access Kafka, see xref:proc-accessing-kafka-using-nodeports-{context}[].

ifdef::Kubernetes[]
=== Exposing Kafka using {KubernetesName} `Ingress`

An external listener of type `ingress` exposes Kafka by using {KubernetesName} `Ingress` and the {NginxIngressController}.
A dedicated `Ingress` resource is created for every Kafka broker pod.
An additional `Ingress` resource is created to serve as a Kafka bootstrap address.
Kafka clients can use these `Ingress` resources to connect to Kafka on port 443.

NOTE: External listeners using `Ingress` have been currently tested only with the {NginxIngressController}.

> {ProductName} is using the TLS passthrough feature of the {NginxIngressController}.

Reviewer 2: ```suggestion
{ProductName} uses the TLS passthrough feature of the {NginxIngressController}.
```

Make sure TLS passthrough is enabled in your {NginxIngressController} deployment.

> For more information abotu enabling TLS passthrough see {NginxIngressControllerTLSPassthrough}.

Reviewer 3: ```suggestion
For more information about enabling TLS passthrough see {NginxIngressControllerTLSPassthrough}.
```

Reviewer 2: ```suggestion
{ProductName} uses the TLS passthrough feature of the {NginxIngressController}.
```

> Because it is using the TLS passthrough functionality, TLS encryption will be always used when exposing Kafka using `Ingress`.

Reviewer 4: just a suggestion ... but maybe it's only me, I would add " ... TLS encryption on the Kafka side ..." in order to be more precise to explain that TLS passthrough means that Nginx will not handle TLS but in any case on Kafka you have to have it enable because the TLS handshake will happen there.

Author: Does it really matter who actually handles it? You still have to use the TLS. The main point of this is to make it clear that you cannot disable TLS because of the TLS passthrough. Should it be maybe changed that way? 

_Because it is using the TLS passthrough functionality, TLS encryption cannot be disbled when exposing Kafka using `Ingress`._

Reviewer 4: Sounds better ... even if I like to be explicit with the reader to drive him understanding that anyway the TLS will be handled by Kafka. But if you like this way, no need to change it's really a nit.

The Ingress controller does not assign any hostnames automatically.

> You have to specify the hostnames which should be used by the bootstrap and per-broker services in the `configuration` section.

Reviewer 3: Can we fully qualify the `configuration` section?

You also have to make sure that the hostnames resolve to the Ingress endpoints.

> {ProductName} will not perform any validation that the requested hosts are available and properly routed to the Ingress endpoints.

Reviewer 3: I suppose we _could_ do this, if we thought it was worth the effort.

Author: That would be quite complicated ... we would need to setup our own backend, setup a test ingress for it and check that we can reach it. And that alone might still not work if there are some network restrictions, DNS limitations etc. So I would probably say that we could do it, but it would be very complicated.
